File: frontend/src/types/skills.ts
/**
 * Adaptive Learning Engine types — mirror backend Pydantic schemas in
 * backend/app/schemas/skills.py
 */

export interface SkillNodeResponse {
  name: string
  domain: string
  description: string
  difficulty_tier: number
}

export interface SkillEdgeResponse {
  from_skill: string
  to_skill: string
  base_weight_hours: number
  adjusted_weight_hours: number | null
  domain: string | null
}

export interface CompanyProfile {
  id: string
  company: string
  role: string
  required_skills: string[]
  preferred_skills: string[]
  difficulty_tier: number
}

export interface MasteryItem {
  skill: string
  mastery: number
}

export interface MasterySnapshotResponse {
  declared_skills: string[]
  quiz_topics: Record<string, number>
  per_skill: MasteryItem[]
}

export interface GapBreakdownResponse {
  student_has: string[]
  missing: string[]
  present_count: number
  required_count: number
  coverage_percent: number
  gap_percent: number
  cosine_similarity: number
}

export interface PathEdge {
  from_skill: string
  to_skill: string
  weight: number
}

export interface SkillPathResult {
  target: string
  total_hours: number
  nodes: string[]
  edges: PathEdge[]
  reachable: boolean
}

export interface MSTEdge {
  from_skill: string
  to_skill: string
  weight: number
}

export interface MSTResponse {
  nodes: string[]
  edges: MSTEdge[]
  total_weight: number
}

export interface KnapsackItemResponse {
  topic: string
  hours: number
  value: number
  reason: string
  included: boolean
}

export interface KnapsackResponse {
  budget_hours: number
  total_hours: number
  total_value: number
  predicted_improvement_percent: number
  items: KnapsackItemResponse[]
}

export interface AdaptivePlanRequest {
  company: string
  role?: string | null
  available_hours: number
}

export interface AdaptivePlanResponse {
  plan_id: string | null
  company: CompanyProfile
  mastery: MasterySnapshotResponse
  gap: GapBreakdownResponse
  paths: SkillPathResult[]
  mst: MSTResponse
  knapsack: KnapsackResponse
}

export interface StudyOptimizerHistoryRow {
  id: string
  target_company: string | null
  target_role: string | null
  available_hours: number
  included_topics: Record<string, unknown>[]
  predicted_improvement: number
  actual_improvement: number | null
  created_at: string
}

File: backend/app/schemas/skills.py
"""Pydantic schemas for the Adaptive Learning Engine (Phase 15)."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class SkillNodeResponse(BaseModel):
    name: str
    domain: str
    description: str
    difficulty_tier: int


class SkillEdgeResponse(BaseModel):
    from_skill: str
    to_skill: str
    base_weight_hours: float
    adjusted_weight_hours: float | None = None  # populated in dynamic responses
    domain: str | None = None


class CompanyProfile(BaseModel):
    id: uuid.UUID
    company: str
    role: str
    required_skills: list[str]
    preferred_skills: list[str]
    difficulty_tier: int


class MasteryItem(BaseModel):
    skill: str
    mastery: float


class MasterySnapshotResponse(BaseModel):
    declared_skills: list[str]
    quiz_topics: dict[str, float]
    per_skill: list[MasteryItem]


class GapBreakdownResponse(BaseModel):
    student_has: list[str]
    missing: list[str]
    present_count: int
    required_count: int
    coverage_percent: float
    gap_percent: float
    cosine_similarity: float


class PathEdge(BaseModel):
    from_skill: str
    to_skill: str
    weight: float


class SkillPathResult(BaseModel):
    target: str
    total_hours: float
    nodes: list[str]
    edges: list[PathEdge]
    reachable: bool


class MSTEdge(BaseModel):
    from_skill: str
    to_skill: str
    weight: float


class MSTResponse(BaseModel):
    nodes: list[str]
    edges: list[MSTEdge]
    total_weight: float


class KnapsackItemResponse(BaseModel):
    topic: str
    hours: float
    value: float
    reason: str
    included: bool


class KnapsackResponse(BaseModel):
    budget_hours: float
    total_hours: float
    total_value: float
    predicted_improvement_percent: float
    items: list[KnapsackItemResponse]


class AdaptivePlanRequest(BaseModel):
    company: str = Field(..., min_length=1, max_length=100)
    role: str | None = None
    available_hours: float = Field(..., ge=1, le=80)


class AdaptivePlanResponse(BaseModel):
    plan_id: uuid.UUID | None
    company: CompanyProfile
    mastery: MasterySnapshotResponse
    gap: GapBreakdownResponse
    paths: list[SkillPathResult]
    mst: MSTResponse
    knapsack: KnapsackResponse


class StudyOptimizerHistoryRow(BaseModel):
    id: uuid.UUID
    target_company: str | None
    target_role: str | None
    available_hours: float
    included_topics: list[dict]
    predicted_improvement: float
    actual_improvement: float | None
    created_at: datetime
